av1an-core/src/grain.rs

```diff
@@ -0,0 +1,161 @@
+use std::{
+  fs::File,
+  io::{BufWriter, Write},
+  path::Path,
+};
+
+#[derive(Debug, Clone, Copy)]
+pub struct NoiseGenArgs {
+  pub iso_setting: u32,
+  pub width: u32,
+  pub height: u32,
+  pub transfer_function: TransferFunction,
+}
+
+const NUM_Y_POINTS: usize = 14;
+type ScalingPoints = [[u8; 2]; NUM_Y_POINTS];
+
+#[allow(clippy::upper_case_acronyms)]
+#[derive(Debug, Clone, Copy)]
+pub enum TransferFunction {
+  /// For SDR content
+  BT1886,
+  /// For HDR content
+  SMPTE2084,
+}
+
+const PQ_M1: f32 = 2610. / 16384.;
+const PQ_M2: f32 = 128. * 2523. / 4096.;
+const PQ_C1: f32 = 3424. / 4096.;
+const PQ_C2: f32 = 32. * 2413. / 4096.;
+const PQ_C3: f32 = 32. * 2392. / 4096.;
+
+impl TransferFunction {
+  pub fn to_linear(self, x: f32) -> f32 {
+    match self {
+      TransferFunction::BT1886 => x.powf(2.8),
+      TransferFunction::SMPTE2084 => {
+        let pq_pow_inv_m2 = x.powf(1. / PQ_M2);
+        (0f32.max(pq_pow_inv_m2 - PQ_C1) / (PQ_C2 - PQ_C3 * pq_pow_inv_m2)).powf(1. / PQ_M1)
+      }
+    }
+  }
+
+  #[allow(clippy::wrong_self_convention)]
+  pub fn from_linear(self, x: f32) -> f32 {
+    match self {
+      TransferFunction::BT1886 => x.powf(1. / 2.8),
+      TransferFunction::SMPTE2084 => {
+        let linear_pow_m1 = x.powf(PQ_M1);
+        (PQ_C2.mul_add(linear_pow_m1, PQ_C1) / PQ_C3.mul_add(linear_pow_m1, 1.)).powf(PQ_M2)
+      }
+    }
+  }
+
+  pub fn mid_tone(self) -> f32 {
+    match self {
+      TransferFunction::BT1886 => 0.18,
+      TransferFunction::SMPTE2084 => 26. / 10000.,
+    }
+  }
+}
+
+fn generate_photon_noise(args: NoiseGenArgs) -> ScalingPoints {
+  // Assumes a daylight-like spectrum.
+  // https://www.strollswithmydog.com/effective-quantum-efficiency-of-sensor/#:~:text=11%2C260%20photons/um%5E2/lx-s
+  const PHOTONS_PER_SQ_MICRON_PER_LUX_SECOND: f32 = 11260.;
+
+  // Order of magnitude for cameras in the 2010-2020 decade, taking the CFA into account.
+  const EFFECTIVE_QUANTUM_EFFICIENCY: f32 = 0.2;
+
+  // Also reasonable values for current cameras. The read noise is typically
+  // higher than this at low ISO settings but it matters less there.
+  const PHOTO_RESPONSE_NON_UNIFORMITY: f32 = 0.005;
+  const INPUT_REFERRED_READ_NOISE: f32 = 1.5;
+
+  // Focal plane exposure for a mid-tone (typically a 18% reflectance card), in lx·s.
+  let mid_tone_exposure = 10. / args.iso_setting as f32;
+
+  // Assumes a 35mm sensor (36mm × 24mm).
```

Comment:
Would it be reasonable to allow a configurable sensor size for more fine-tuning?

Reply:
Yes, as it would allow for large grains, which is one weakness of how grain synthesis normally works.

It can look wonky however, which is why normal grain synth+better grain preservation encoder side wise is better.

Reply:
What would this configuration look like? I'm asking because I'm certainly no expert on camera sensors. This math was done by whoever wrote the photon noise tool in aomenc's repo.

Reply:
I know just the right person for this, but I don't think he'll be free for a little while.

Reply:
Are we talking days, or weeks? If it's more than a couple of days, I'd move to go forward with this PR for now and make an issue to follow up later to add this configuration option.

Reply:
Let's tweak this later then and move forward with the PR.

Reply:
Days, no real worries.

In the worst case scenario, we can just do what you said Soichiro.